traits/trait_list_object.py

```diff
@@ -53,465 +90,765 @@ def __init__(self, index=0, removed=None, added=None):
-    appropriate index, added and removed values.  In the case of setting
-    or deleting items with a slice, the index will hold either:
-
-    - the location of the first element changed, if the step is 1.  In
-      this case the last element changed can be inferred from the
-      length of the removed values.
-    - the slice that was used, otherwise.
+class TraitList(list):
+    """ A subclass of list that validates and notifies listeners of changes.
 
     Parameters
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : HasTraits instance
-        The HasTraits instance that the list has been set as an attribute for.
-    name : str
-        The name of the attribute on the object.
     value : list
-        The list of values to initialize the TraitListObject with.
+        The value for the list
+    validator : callable, optional
+        Called to validate items in the list.
+        With the signature::
+
+            validator(current_list, index, removed, added)
+
+        If the validator is not provided, all values added to the list
+        are considered to be valid.
+
+    notifiers : list of callable, optional
+        A list of callables with the signature::
+
+            notifier(trait_list, index, removed, added)
+
+        If a notifier is not provided, an empty list of notifiers is
+        used.
 
     Attributes
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : weak reference to a HasTraits instance
-        A weak reference to a HasTraits instance that the list has been set
-        as an attribute for.
-    name : str
-        The name of the attribute on the object.
-    name_items : str
-        The name of the items event trait that the trait list will fire when
-        mutated.
+    validator : callable, optional
+        Called to validate items in the list
+        With the signature::
+
+            validator(current_list, index, removed, added)
+
+        If the validator is not provided, all values added to the list
+        are considered to be valid.
+
+    notifiers : list of callable, optional
+        A list of callables with the signature::
+
+            notifier(trait_list, index, removed, added)
+
+        If a notifier is not provided, an empty list of notifiers is
+        used.
+
     """
 
-    def __init__(self, trait, object, name, value):
-        self.trait = trait
-        self.object = ref(object)
-        self.name = name
-        self.name_items = None
-        if trait.has_items:
-            self.name_items = name + "_items"
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls)
+        instance.validator = None
+        instance.notifiers = []
+        return instance
+
+    def __init__(self, value=None, *, validator=None, notifiers=()):
```

Comment:
Sorry; I think there was a miscommunication. It's the default for `notifiers` that I want to change here. `value` is a general iterable, so it's fine for that to be `()` (and I think that's preferable, because it matches the signature for the first argument for Python's built-in `list`). But `notifiers` is a list, so it's misleading to use a tuple as the default. Instead, we should follow the usual practice for mutable defaults.

Please:

- restore the default for `value` to `()`. (Otherwise we're potentially passing `None` to our general-purpose `validate` method.)
- change the default for `notifiers` to `None`
- do an explicit `if notifiers is None` check within the method so that `notifiers` is treated properly as an empty list in that case.

Actually, I think we have another issue here, and one that may be worth an extra unit test: `self.validate` expects a list, but for compatibility with the `list` base class, `value` is allowed to be a general iterable. So we probably want to explicitly convert `value` to a list at some point before validating.

Reply:
> restore the default for value to (). (Otherwise we're potentially passing None to our general-purpose validate method.)
> value is allowed to be a general iterable.

Yes I just spotted these in `TraitSet` and came back here to find the same thing... Sorry for not having spotted this earlier.

Reply:
I'm sorry, my mistake, changed default `value` back, added none check for  `notifiers` and converting `value` to list before base validation. Also added a unittest for an iterable `value`.

Reply:
> Yes I just spotted these in `TraitSet` and came back here to find the same thing... Sorry for not having spotted this earlier.

Yep, fixing that there.